### ISSUE TITLE
Apply a mixin to a class that also directly implements the "super-class constraints"

### ISSUE THREAD

**Issue by rich-j** [dart-lang/sdk#35202]
Dart 2.1 supports `mixin` with "super-class constraints" such as:

    abstract class S {
      String get str;
    }

    mixin M on S {
      void f() { print(str); }
    }

To apply the mixin to a class requires the class to "implement" the "`on` type requirements of the mixin declaration". The following declaration:

    class A with M implements S {   // error: The class doesn't implement the required class 'S.
      final String str = "In A";
    }

results in an analyzer error that `The class doesn't implement the required class 'S'`.

A declaration that does work uses an (effectively "no-op") abstract class that declares the `implements` (but doesn't need to implement it because it's abstract):

    abstract class _BX implements S {}

    class B extends _BX with M {
      final String str = "In B";
    }

I would like to see a class declaration (such as `class A` above) that combines the `with` and `implements` clauses as a valid application of a mixin with super-class constraints and not need to provide an extra layer of abstraction to declare the `implements`.

**Comment by lrhn** [comment]
The `on` clause of the `mixin` declaration is there to ensure that you can make *super* invocations. It's a requirement on the super-class of the mixin application.

You don't actually *need* the mixin application super-class to be an `S` here, you just need the class itself to be an `S`. You can make that requirement by declaring `M` as:
```dart
mixin M implements S {
  void f() { print(str); }
}
```
This way, any mixin-application of `M` must provide a full implementation of `S`. The super-class doesn't have to *implement* `S` itself, but the super-class, the mixin and the class itself must provide implementations of all the members of `S`.

(In this particular degenerate example, you an also just write;
```dart
class A extends S with M {
  final String str = "In A";
}
```
It may not always be that simple).

So, in short: Don't use an `on` clause unless you actually need to make super-invocations on the super-class.

**Comment by rich-j** [comment]
Thank you - yes, the `implements` clause is what I need. I usually think of `implements` as "provides", but in an abstract declaration (such as a mixin) it can also be a "requires".

Your above examples would be good in the [Language Tour - ...mixins](https://www.dartlang.org/guides/language/language-tour#adding-features-to-a-class-mixins) where it says:

> To specify that only certain types can use the mixin — for example, so your mixin can invoke a method that it doesn’t define — use `on` to specify the required superclass:

This led me to think that `on` is how to specify a general "requires".

**Comment by hacker1024** [comment]
Would using `implements` instead of `on` be appropriate here? This forces classes that mix in the mixin to implement the implemented class, without the `super` issue.
